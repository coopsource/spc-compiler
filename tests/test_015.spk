define function output taking arguments tape and length and pointer as
    print the string Tape contents
    new line
    set I to 1
    while I < length+1 do
        set X to element I of tape + 0
		print X
        set I to I + 1
    end while
    new line
    set I to 1
    while I < pointer do
		print space
		set I to I + 1
	end while
	print the string ^
	new line
end function

define function main taking no arguments as
	read N
	read M
	read Q
	set I to 0
	set done to 0
	set S to N * N 
	set S to S + M
	set element S of update to 0
	set element S of move to 0
	set element S of next to 0
	while done = 0 do
	    read state
	    read symbol
	    if state = 0 and symbol = 0 then
            set done to 1
	    else
		    set row to state - 1
			set index to row * N + symbol
		    read element index of update
		    read element index of move
		    read element index of next
			set I to I + 1
		end if
	end while

	set element 1 of tape to 0
	set I to 1
	set done to 0
	while done = 0 do
		read T
		if T = 0 then
		    set done to 1
		else
			set element I of tape to T
			set I to I + 1
		end if
	end while
		
	set max to I - 1
	set pointer to 1
	set state to Q
	set done to 0
	set step to 1
	while done = 0 do
		set row to state - 1
		set col to element pointer of tape
		set index to row * N + col
		print the string Step
		print space
		print step
		new line
		print the string State
		print space
		print state
		new line
		call output with tape and max and pointer
		set element pointer of tape to element index of update
		set pointer to pointer + element index of move
		set state to element index of next
	    if pointer > max then
	    	set element pointer of tape to 0
			set max to pointer
		else
			nothing
		end if
		if state = 0 then
			set done to 1
		else
			nothing
		end if
		set step to step + 1
	end while
	
	print the string Program completed
	new line
	call output with tape and max and pointer
end function
