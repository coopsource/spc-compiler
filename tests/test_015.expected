Step 1
State 1
Tape contents
112
^
Step 2
State 1
Tape contents
212
 ^
Step 3
State 1
Tape contents
222
  ^
Program completed
Tape contents
2210
   ^
